book/src/error_handling.md

```diff
@@ -31,27 +49,116 @@ pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_divide(
         result
     } else {
         // Divisor is zero -- throw an exception!
-        env.throw_new("java/lang/ArithmeticException", "Attempting to divide by zero.");
-        // It doesn't matter what we return now, since the JNI will see that an
-        // exception was thrown.
+        env.throw_new("java/lang/Exception", "Attempting to divide by zero.");
+        // A dummy value must be returned, since the native method retains control
+        // even after throwing an Exception.
         0
     }
 }
 ```
+    
+## Wrapping Result Handling into a Function
+
+It can be helpful to wrap common error handling into one function, so that there
+is less repeated JNI code. For ease of implementation, we'll use
+`anyhow::Error`, and make the caller pass the `dummy_value`, which is returned
+only on error paths. The `JNIEnv` APIs all throw `Result<T,E>`, and the
+`JavaException` variant marks that an exception has already been thrown - so for
+that variant alone, we will skip throwing an Exception. You may come up with
+your own Error hierarchies, which could have a different logic.
+
+```rust
+fn try_java<F, T>(env: JNIEnv, dummy_value: T, f: F) -> T
+where
+    F: FnOnce() -> Result<T, anyhow::Error>,
+{
+    match f() {
+        Ok(s) => s,
+        Err(e) => {
+            if let Some(e) = e.downcast_ref::<jni::errors::Error>() {
+                match e {
+                    // Since JavaException implies an Exception has already been thrown, 
+                    // don't throw another one.
+                    jni::errors::Error::JavaException => {},
+                    _ => {
+                        let _ = env.throw_new("java/lang/Exception", e.to_string());
```

Comment:
It is not enough to prevent exception masking: you might get some error type that is not `jni::errors::Error::JavaException` _and_ have a pending exception in Java. The only reliable way is to check if there is an exception pending already:

https://github.com/exonum/exonum-java-binding/blob/a6627aa00d10bc8175d68392e9156fbf99f48bfe/exonum-java-binding/core/rust/src/utils/errors.rs#L162

Reply:
Possibly, even before downcast (see the comment below)

Reply:
I think I see your point - if we have some nested calls that translate a `JavaException` into some other value or type, this will fail.

Reply:
Possibly; we can expect anything from BL code :-) 

Also, just as with the module that controls the longs flying between Java and Rust to be valid pointers to objects of correct types (aka resource-manager); this method might be needed to many projects. If we can come up with a reasonably generic and performant one, we can include it in the library, as a separate PR.